R/shuffle_ard.R

```diff
@@ -0,0 +1,216 @@
+#'   ard_categorical(ADSL, variables = "ARM")
+#' ) |>
+#' shuffle_ard()
+#'
+shuffle_ard <- function(x, trim = TRUE){
+
+  check_class_data_frame(x)
+
+  # make sure columns are in order & add index for retaining order
+  dat_cards <- x |>
+    dplyr::relocate(dplyr::starts_with("group"), dplyr::starts_with("variable"),
+                    dplyr::any_of(c("stat_name", "stat_label", "statistic")), .before = 0L) %>%
+    dplyr::arrange(dplyr::pick(dplyr::any_of("variable")),
+                   dplyr::pick(dplyr::matches("^group[0-9]+$")),
+                   dplyr::pick(dplyr::matches("^group[0-9]+_level$"))) |>
+    dplyr::mutate(.cards_idx = dplyr::row_number())
+
+  # fill stat label if missing
+  dat_cards <- dat_cards |>
+    dplyr::mutate(dplyr::across(any_of("stat_label"), ~ dplyr::coalesce(.x, stat_name)))
+
+  # split up the data into data/variable info & cards info
+  vars_ard <- dat_cards |> dplyr::select(all_ard_groups(), all_ard_variables()) |> names()
+  vars_protected <- setdiff(names(dat_cards), vars_ard)
+
+  dat_cards_grps <- dat_cards |>
+    dplyr::select(-all_of(vars_protected), ".cards_idx")
+
+  dat_cards_stats <- dat_cards |>
+    dplyr::select(all_of(vars_protected))
+
+  # process the data/variable info
+  dat_cards_grps_processed <- dat_cards_grps |>
+    # unlist the list-columns
+    dplyr::mutate(
+      dplyr::across(
+        where(.is_list_column_of_scalars),
+        ~lapply(., \(x) if (!is.null(x)) x else NA_character_) |> unlist()
+      )
+    ) |>
+    .check_var_nms(vars_protected = names(dat_cards_stats)) |>
+    .rnm_grp_vars() |>
+    .fill_grps_from_variables()
+
+  # join together again
+  dat_cards_out <- dplyr::left_join(
+    dat_cards_grps_processed,
+    dat_cards_stats,
+    by = ".cards_idx"
+  )
+
+  # fill stat_label -> variable_level if exists
+  if ("stat_label" %in% names(dat_cards_out)){
+    dat_cards_out <- dat_cards_out |>
+      dplyr::mutate(dplyr::across(any_of("variable_level"), ~dplyr::coalesce(.x, stat_label)))
+  }
+
+  dat_cards_out <- dat_cards_out |>
+    dplyr::rename(any_of(c(label="variable_level"))) |>
+    dplyr::arrange(".cards_idx")   |>
+    dplyr::select(-".cards_idx")
+
+  if (trim){
+    dat_cards_out |> .trim_ard()
+  } else {
+    dat_cards_out
+  }
+}
+
+
+# trim extra variables and subset to numeric results only
+.trim_ard <- function(x) {
+
+  check_class_data_frame(x)
+
+  # flatten ard table for easier viewing ---------------------------------------
+  x |>
+    # remove the formatting functions / warning / error
+    dplyr::select(-where(function(x) all(lapply(x, function(y) is.null(y) || is.function(y)) |> unlist())),
+                  -"warning", -"error",
+                  -any_of(c("stat_label"))) |>
+    # filter to numeric statistic values
+    dplyr::filter(map_lgl(.data$statistic, is.numeric)) |>
+    # unlist the list-columns
+    dplyr::mutate(
+      dplyr::across(
+        where(.is_list_column_of_scalars),
```

Comment:
I think the definition of `.is_list_column_of_scalars()` should be more strict:
1. We only want this to flatten the statistics column, right? Should we just specify the column, instead of the `where()`?
2. The function is picking up on the warning and error columns, but this will not always be the case. For example, if a function returns two warnings, the text of both warnings are saved as a vector. It's probably best that the trimming returns the same columns regardless if there were 0, 1, or 2+ warnings thrown.

Reply:
So I originally wrote this so it could be used standalone on a cards object, in which case you may end up with more list-columns (like group_level, if some values are character and some are NULL). But since we are making the function internal within the shuffle_ard() call, this makes perfect sense!